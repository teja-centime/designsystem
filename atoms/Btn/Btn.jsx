import React from "react";
import PropTypes from "prop-types";
import Button from "@material-ui/core/Button";
import BtnLoader from "atm/src/atoms/BtnLoader/BtnLoader";
import {useT} from "atm/src/utils/Translation/centimeTrans";
import {HeightClasses} from "atm/src/utils/Global/cssConstants";

export const BtnColors = {
    primary: "primary",
    secondary: "secondary",
    error: "error",
};
export const BtnVariants = {
    text: "text",
    outlined: "outlined",
    contained: "contained",
};

let BtnSizes = {...HeightClasses};
delete BtnSizes.Large40;

const MaterialBtnSizes = {
    [HeightClasses.ExtraLarge44]: "large",
    [HeightClasses.Medium36]: "medium",
    [HeightClasses.Small32]: "small",
};

export const BtnPropTypes = {
    /**
     *  Text to be displayed in the button
     */
    label: PropTypes.string,
    /**
     * Various Colors of a Button we have :
     *
     */
    color: PropTypes.oneOf(Object.values(BtnColors)),
    /**
     * Variants of each color button with three types used in different scenarios
     * Check out for nice usages of each type : https://material.io/components/buttons/#text-button
     *
     * @default 'contained'
     */
    variant: PropTypes.oneOf(Object.values(BtnVariants)),
    /**
     * Sizes of the button
     * Large has 3rem height; normal has 2.75rem height; small has 2.5rem height
     */
    size: PropTypes.oneOf([...Object.values(BtnSizes)]),
    /**
     *
     *  Spans the full width of the Button parent
     *
     *  */
    fullWidth: PropTypes.bool,
    /**
     * Disables the Button, preventing mouse events,
     * even if the underlying component is an `<a>` element
     */
    disabled: PropTypes.bool,
    /**
     * Providing a `href` will render an `<a>` element, _styled_ as a button.
     * */
    href: PropTypes.string,
    /**
     * Icon with the Button
     */
    icon: PropTypes.node,
    /**
     * If this button only contains the icon
     */
    isIconBtn: PropTypes.bool,
    /**
     * Icon Position
     */
    iconPosition: PropTypes.oneOf(["left", "right"]),
    /**
     * Defines HTML button type attribute.
     *
     * @default 'button'
     */
    type: PropTypes.oneOf(["button", "reset", "submit", null]),

    /**
     * Can Render the button as `button` html element or `a` element or `input` etc..
     *
     * @default 'button'
     */
    as: PropTypes.elementType,

    /**
     * Class Names space separated.
     */
    className: PropTypes.string,

    /**
     * On Click Function
     */
    onClick: PropTypes.func,
    /**
     * Loading State of a Button
     */
    loading: PropTypes.bool,
    /**
     * Translation File's NameSpace - TNS - Can be a string or an array
     */
    tns: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
    /**
     * Id of the component
     */
    id: PropTypes.string,
};

/**
 *
 * Usage : `import Btn from "atm/src/atoms/Btn/Btn"`
 */
const Btn = React.forwardRef(
    (
        {
            label,
            color = BtnColors.primary,
            variant = BtnVariants.contained,
            size = HeightClasses.ExtraLarge44,
            fullWidth = false,
            disabled = false,
            href = null,
            icon,
            isIconBtn,
            iconPosition = "left",
            type = "button",
            as = "button",
            className,
            onClick = null,
            loading = false,
            tns,
            id,
        },
        ref
    ) => {
        return (
            <Button
                disableRipple
                ref={ref}
                color={color}
                variant={variant}
                fullWidth={fullWidth}
                disabled={loading ? true : disabled}
                href={href}
                type={type}
                component={as}
                size={
                    MaterialBtnSizes[size] ? MaterialBtnSizes[size] : undefined
                }
                className={
                    (loading ? "noOpacity " : "") +
                    " " +
                    (color === BtnColors.error ? "btnError " : "") +
                    (isIconBtn ? "iconBtn " : "") +
                    (className ? className : "") +
                    (size ? " " + size : "")
                }
                onClick={onClick}
                id={id || label}
            >
                {loading ? (
                    <div className={"btnLoader"}>
                        <BtnLoader/>{" "}
                    </div>
                ) : (
                    <>
                        {iconPosition === "left" && icon && icon}
                        {useT(label, tns)}
                        {iconPosition === "right" && icon && icon}
                    </>
                )}
            </Button>
        );
    }
);

Btn.propTypes = BtnPropTypes;

export default Btn;
