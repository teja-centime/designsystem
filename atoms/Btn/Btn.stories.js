import React from "react";
import Btn, { BtnColors, BtnVariants } from "./Btn";
import { withA11y } from "@storybook/addon-a11y";
import Icon from "atm/src/atoms/Icon/Icon";
import { HeightClasses } from "atm/src/utils/Global/cssConstants";
// import Icon from "../Icon/Icon";

export default {
    component: Btn,
    title: "Atoms/Button",
    parameters: {
        component: Btn,
        componentSubtitle: "All Types Buttons in Centime.",
        "in-dsm": { id: "5f97e679100ed33930f3b2c3", version: "0.0.2" },
        decorators: [withA11y],
        actions: {
            argTypesRegex: "^on.*",
            handles: ["mouseover", "click .btn"],
        },
    },
    argTypes: {
        color: {
            control: {
                type: "inline-radio",
                options: BtnColors,
            },
        },
        variant: {
            control: {
                type: "inline-radio",
                options: BtnVariants,
            },
        },
        as: {
            control: {
                type: "inline-radio",
                options: ["a", "input", "button"],
            },
        },
        type: {
            control: {
                type: "inline-radio",
                options: ["reset", "submit", "button"],
            },
        },
        iconPosition: {
            control: {
                type: "inline-radio",
                options: ["left", "right"],
            },
        },
        onClick: {
            action: "clicked",
        },
    },
};

const BtnTemplate = args => <Btn {...args} />;

export const all = args => {
    return (
        <>
            <Btn
                onClick={() => {}}
                label="Primary"
                color="primary"
                size={HeightClasses.ExtraLarge44}
            />
            &nbsp;
            <Btn
                onClick={() => {}}
                label="Secondary"
                color="secondary"
                size={HeightClasses.ExtraLarge44}
            />
            &nbsp;
            <Btn
                onClick={() => {}}
                label="Outlined Secondary"
                color="secondary"
                variant="outlined"
                size={HeightClasses.ExtraLarge44}
            />
            &nbsp;
            <Btn
                onClick={() => {}}
                label="Text Primary"
                color="primary"
                variant={"text"}
                size={HeightClasses.ExtraLarge44}
            />
            &nbsp;
            <Btn
                onClick={() => {}}
                label="Text Secondary"
                color="secondary"
                variant={"text"}
                size={HeightClasses.ExtraLarge44}
            />
            <br />
            <br />
            <Btn
                onClick={() => {}}
                label="Disabled"
                disabled={true}
                size={HeightClasses.Medium36}
            />
            &nbsp;
            <Btn
                onClick={() => {}}
                label="Disabled"
                color={"secondary"}
                disabled={true}
                size={HeightClasses.Medium36}
            />
            &nbsp;
            <Btn
                onClick={() => {}}
                label="Disabled Outline"
                color={"secondary"}
                variant={"outlined"}
                disabled={true}
                size={HeightClasses.Medium36}
            />
            &nbsp;
            <br />
            <br />
            <Btn
                onClick={() => {}}
                label="Default Height - Large"
                color={"secondary"}
                variant={"outlined"}
                size={HeightClasses.ExtraLarge44}
            />
            &nbsp;
            <Btn
                onClick={() => {}}
                label="Medium Height"
                color={BtnColors.secondary}
                size={HeightClasses.Medium36}
                variant={BtnVariants.outlined}
            />
            &nbsp;
            <Btn
                onClick={() => {}}
                label="Small Height"
                color={BtnColors.secondary}
                size={HeightClasses.Small32}
                variant={BtnVariants.outlined}
            />
            &nbsp;
            <br />
            <br />
            <div style={{ width: "25%" }}>
                <Btn
                    onClick={() => {}}
                    label="Full Width Primary"
                    color="primary"
                    fullWidth={true}
                />
            </div>
            <br />
            <div style={{ width: "25%" }}>
                <Btn
                    onClick={() => {}}
                    label="Loading"
                    loading={true}
                    fullWidth={true}
                />
            </div>
            <br />
            <div style={{ width: "25%" }}>
                <Btn
                    onClick={() => {}}
                    loading={true}
                    fullWidth={true}
                    label="Secondary"
                    color="secondary"
                />
            </div>
            <br />
            <div style={{ width: "25%" }}>
                <Btn
                    onClick={() => {}}
                    loading={true}
                    fullWidth={true}
                    label="Outlined Secondary"
                    color="secondary"
                    variant="outlined"
                />
            </div>
            <br />
            <Btn onClick={() => {}} label="Loading Small Btn" loading={true} />
            <br />
            <br />
            <Btn
                onClick={() => {}}
                label="Remove"
                color={BtnColors.secondary}
                icon={
                    <Icon
                        name={"delete"}
                        customFontSize={"0.75rem"}
                        className={"mr-1"}
                    />
                }
                size={HeightClasses.Medium36}
                variant={"outlined"}
            />
            &nbsp;
            <Btn
                onClick={() => {}}
                label="Reconnect"
                color={BtnColors.error}
                icon={<Icon name="thunder" color={"error"} />}
                size={HeightClasses.Medium36}
                variant={"outlined"}
            />
            &nbsp;
            <Btn
                onClick={() => {}}
                label="Download"
                icon={
                    <Icon
                        name="download"
                        customFontSize={"0.75rem"}
                        className={"mr-2"}
                    />
                }
                size={HeightClasses.Medium36}
                color="secondary"
                variant={"text"}
            />
            &nbsp;
            <br />
            <br />
            <p>Pagination Buttons</p>
            <Btn
                onClick={() => {}}
                label=""
                color={BtnColors.secondary}
                icon={<Icon name="back" />}
                size={HeightClasses.Medium36}
                isIconBtn
                variant={"outlined"}
                className={
                    "noBorderTopRightRadius noBorderBottomRightRadius leftRoundedBtn"
                }
            />
            <Btn
                onClick={() => {}}
                label=""
                color={BtnColors.secondary}
                icon={<Icon name="forward" />}
                size={HeightClasses.Medium36}
                isIconBtn
                className={"noBorderTopLeftRadius noBorderBottomLeftRadius"}
                variant={"outlined"}
            />
            <br />
            <br />
            <p>Carousel Buttons</p>
            <Btn
                onClick={() => {}}
                label=""
                color={BtnColors.secondary}
                icon={
                    <Icon
                        name="less-than"
                        className={"mb-1"}
                        customFontSize={"0.5rem"}
                    />
                }
                size={HeightClasses.Small32}
                isIconBtn
                variant={"outlined"}
                className={"mr-2"}
            />
            <Btn
                onClick={() => {}}
                label=""
                color={BtnColors.secondary}
                icon={
                    <Icon
                        name="greater-than"
                        className={"mb-1"}
                        customFontSize={"0.5rem"}
                    />
                }
                size={HeightClasses.Small32}
                isIconBtn
                variant={"outlined"}
            />
        </>
    );
};
all.args = {};

all.story = {
    parameters: {
        "in-dsm": {
            id: "5f994fddbf85ef80c43d1b10",
        },
    },
};

export const PrimaryButton = BtnTemplate.bind({});
PrimaryButton.args = {
    color: "primary",
    label: "Primary Button",
};
PrimaryButton.story = {
    parameters: {
        // Story-Level 'in-dsm' configuration (Will apply only to the story that is being configured)
        // Story-Level configuration will override Module-Level 'in-dsm' configuration for the specific story
        // "in-dsm": {
        //     id: "5f97e679100ed33930f3b2c3",
        //     versionFilePath: "../../../package.json"
        // },
    },
};

export const PrimaryText = BtnTemplate.bind({});
PrimaryText.args = {
    ...PrimaryButton.args,
    variant: "text",
};
//
// export const PlayWithProps = args => (
//     <div style={{ width: "25%" }}>
//         <Btn {...args} />
//     </div>
// );
// PlayWithProps.args = {
//     label: "Play with Knobs",
//     disabled: false,
//     color: "primary",
//     variant: "contained",
//     fullWidth: false,
//     as: "button",
//     type: "button",
// };
