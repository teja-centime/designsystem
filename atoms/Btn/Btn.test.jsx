import React, { Suspense } from "react";
import Btn from "./Btn";
import { cleanup, fireEvent, render } from "@testing-library/react";

jest.mock("atm/src/utils/Translation/centimeTrans", () => {
    let originalModule = jest.requireActual("atm/centime.atm.utils.trans");
    let returnModule = Object.assign(originalModule);
    returnModule.useTrans = (props) => (key) => key;
    returnModule.useT = (label, ns = null) => label;
    return returnModule;
});

afterEach(cleanup);
/*describe("Button Rendering..", () =>{
   test('Renders & UnMounts Without Crashing', () =>{
      const div = document.createElement('div');
      ReactDOM.render(<Suspense fallback={"loading"}><Btn label={`Welcome`}/></Suspense>,div);
      ReactDOM.unmountComponentAtNode(div);
   });

});*/

const label = "Welcome";

describe("Button Events - Testing.. ", async () => {
    test("On Click Event", async () => {
        const onClick = jest.fn();
        const { findByText } = render(
            <Suspense fallback={"loading"}>
                <Btn label={label} onClick={onClick} />
            </Suspense>
        );
        const myBtn = await findByText(label);
        fireEvent.click(myBtn);
        expect(onClick).toHaveBeenCalledTimes(1);
    });
});

describe("Button Snapshots - Compare Snapshot of.. ", async () => {
    test("Default Props Button", async () => {
        const { container, findByText } = render(
            <Suspense fallback={"loading"}>
                <Btn label={label} />
            </Suspense>
        );
        await findByText(label);
        expect(container.firstChild).toMatchSnapshot();
    });

    test("Primary Button", async () => {
        const { container, findByText } = render(
            <Suspense fallback={"loading"}>
                <Btn label={label} color={"primary"} />
            </Suspense>
        );
        await findByText(label);
        expect(container.firstChild).toMatchSnapshot();
    });

    test("Secondary Button", async () => {
        const { container, findByText } = render(
            <Suspense fallback={"loading"}>
                <Btn label={label} color={"secondary"} />
            </Suspense>
        );
        await findByText(label);
        expect(container.firstChild).toMatchSnapshot();
    });

    test("Outlined Secondary Button", async () => {
        const { container, findByText } = render(
            <Suspense fallback={"loading"}>
                <Btn label={label} color="secondary" variant="outlined" />
            </Suspense>
        );
        await findByText(label);
        expect(container.firstChild).toMatchSnapshot();
    });

    test("Text Primary Button", async () => {
        const { container, findByText } = render(
            <Suspense fallback={"loading"}>
                <Btn label={label} color="primary" variant={"text"} />
            </Suspense>
        );
        await findByText(label);
        expect(container.firstChild).toMatchSnapshot();
    });

    test("Text Secondary Button", async () => {
        const { container, findByText } = render(
            <Suspense fallback={"loading"}>
                <Btn label={label} color="secondary" variant="text" />
            </Suspense>
        );
        await findByText(label);
        expect(container.firstChild).toMatchSnapshot();
    });

    test("Disabled Primary Button", async () => {
        const { container, findByText } = render(
            <Suspense fallback={"loading"}>
                <Btn label={label} color={"primary"} disabled={true} />
            </Suspense>
        );
        await findByText(label);
        expect(container.firstChild).toMatchSnapshot();
    });

    test("Disabled Secondary Button", async () => {
        const { container, findByText } = render(
            <Suspense fallback={"loading"}>
                <Btn label={label} color={"secondary"} disabled={true} />
            </Suspense>
        );
        await findByText(label);
        expect(container.firstChild).toMatchSnapshot();
    });

    test("Disabled Outlined Secondary Button", async () => {
        const { container, findByText } = render(
            <Suspense fallback={"loading"}>
                <Btn
                    label={label}
                    color="secondary"
                    variant="outlined"
                    disabled={true}
                />
            </Suspense>
        );
        await findByText(label);
        expect(container.firstChild).toMatchSnapshot();
    });

    test("Text Primary Disabled Button", async () => {
        const { container, findByText } = render(
            <Suspense fallback={"loading"}>
                <Btn
                    label={label}
                    color="primary"
                    variant={"text"}
                    disabled={true}
                />
            </Suspense>
        );
        await findByText(label);
        expect(container.firstChild).toMatchSnapshot();
    });

    test("Text Secondary Disabled Button", async () => {
        const { container, findByText } = render(
            <Suspense fallback={"loading"}>
                <Btn
                    label={label}
                    disabled={true}
                    color="secondary"
                    variant="text"
                />
            </Suspense>
        );
        await findByText(label);
        expect(container.firstChild).toMatchSnapshot();
    });

    test("Loading Primary Button", () => {
        const { container } = render(
            <Suspense fallback={"loading"}>
                <Btn label={""} loading={true} color="primary" />
            </Suspense>
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
