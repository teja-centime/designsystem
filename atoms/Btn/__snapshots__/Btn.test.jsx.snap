// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Snapshots - Compare Snapshot of..  Default Props Button 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
  tabindex="0"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    Welcome
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;

exports[`Button Snapshots - Compare Snapshot of..  Disabled Outlined Secondary Button 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary Mui-disabled Mui-disabled"
  disabled=""
  tabindex="-1"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    Welcome
  </span>
</button>
`;

exports[`Button Snapshots - Compare Snapshot of..  Disabled Primary Button 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
  disabled=""
  tabindex="-1"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    Welcome
  </span>
</button>
`;

exports[`Button Snapshots - Compare Snapshot of..  Disabled Secondary Button 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary Mui-disabled Mui-disabled"
  disabled=""
  tabindex="-1"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    Welcome
  </span>
</button>
`;

exports[`Button Snapshots - Compare Snapshot of..  Loading Primary Button 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-contained noOpacity MuiButton-containedPrimary Mui-disabled Mui-disabled"
  disabled=""
  tabindex="-1"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    <div
      class="btnLoader"
    >
      <svg
        class="spinner"
        viewBox="0 0 100 100"
      >
        <defs>
          <lineargradient
            id="Gradient"
            x1="50%"
            x2="50%"
            y1="0%"
            y2="100%"
          >
            <stop
              offset="0%"
              stop-color="#ffffff"
            >
              <animate
                attributeName="stop-color"
                dur="4s"
                repeatCount="indefinite"
                values="#ffffff; #ffb500; #ffffff"
              />
            </stop>
            <stop
              offset="100%"
              stop-color="#ffb500"
            >
              <animate
                attributeName="stop-color"
                dur="4s"
                repeatCount="indefinite"
                values="#ffb500; #ffffff; #ffb500"
              />
            </stop>
          </lineargradient>
        </defs>
        <circle
          class="circle"
          cx="50"
          cy="50"
          fill="none"
          r="30"
        />
      </svg>
       
    </div>
  </span>
</button>
`;

exports[`Button Snapshots - Compare Snapshot of..  Outlined Secondary Button 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary"
  tabindex="0"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    Welcome
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;

exports[`Button Snapshots - Compare Snapshot of..  Primary Button 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
  tabindex="0"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    Welcome
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;

exports[`Button Snapshots - Compare Snapshot of..  Secondary Button 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
  tabindex="0"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    Welcome
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;

exports[`Button Snapshots - Compare Snapshot of..  Text Primary Button 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
  tabindex="0"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    Welcome
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;

exports[`Button Snapshots - Compare Snapshot of..  Text Primary Disabled Button 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled"
  disabled=""
  tabindex="-1"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    Welcome
  </span>
</button>
`;

exports[`Button Snapshots - Compare Snapshot of..  Text Secondary Button 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary"
  tabindex="0"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    Welcome
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;

exports[`Button Snapshots - Compare Snapshot of..  Text Secondary Disabled Button 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary Mui-disabled Mui-disabled"
  disabled=""
  tabindex="-1"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    Welcome
  </span>
</button>
`;
